File: XUCore/additions/NSRecursiveLockAdditions.m
//
//  NSLockAdditions.m
//
//  Created by Charlie Monroe on 2/20/14.
//  Copyright (c) 2014 Charlie Monroe Software. All rights reserved.
//

#import "NSRecursiveLockAdditions.h"

@implementation NSRecursiveLock (Additions)

-(void)performLockedBlock:(void (^)(void))block{
	[self lock];
	
	@try {
		block();
	}@catch (NSException *exception) {
		[self unlock];
		@throw exception;
	}

	[self unlock];
}

@end

File: XUCore/additions/NSRecursiveLockAdditions.swift
//
//  NSRecursiveLockAdditions.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/11/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation

public extension NSRecursiveLock {
	
	public func performLockedBlock(block: (Void) -> Void) {
		self.lock()
		
		let handler = XUExceptionHandler(catchHandler: { (exception) -> Void in
			// We only unlock self if an exception was raised. If no exception
			// occurrs, the lock is unlocked within performing the block.
			self.unlock()
			exception.raise() // Rethrow the exception
		}) { /* No-op finally. */ }
		
		handler.performBlock { () -> Void in
			block()
			self.unlock()
		}
	}
	
	
}
